bot/code_review_bot/sources/phabricator.py

```diff
@@ -0,0 +1,227 @@
+            if not build.revision:
+                raise Exception("Not found")
+
+            # Check against secure projects
+            projects = set(build.revision["attachments"]["projects"]["projectPHIDs"])
+            if projects.intersection(self.secure_projects):
+                raise Exception("Secure revision")
+        except Exception as e:
+            logger.info("Revision not accessible", build=str(build), error=str(e))
+            return False
+
+        return True
+
+    def load_patches_stack(self, build):
+        """
+        Load a stack of patches for a public Phabricator build
+        without hitting a local mercurial repository
+        """
+        build.stack = self.api.load_patches_stack(build.diff_id, build.diff)
+
+    def load_reviewers(self, build):
+        """
+        Load details for reviewers found on a build
+        """
+        assert isinstance(build, PhabricatorBuild)
+        assert build.state == PhabricatorBuildState.Public
+        assert build.revision is not None
+
+        def load_user(phid):
+            if phid.startswith("PHID-USER"):
+                return self.api.load_user(user_phid=phid)
+            elif phid.startswith("PHID-PROJ"):
+                logger.info(f"Skipping group reviewer {phid}")
+            else:
+                raise Exception(f"Unsupported reviewer {phid}")
+
+        reviewers = build.revision["attachments"]["reviewers"]["reviewers"]
+        build.reviewers = list(
+            filter(
+                None, [load_user(reviewer["reviewerPHID"]) for reviewer in reviewers]
+            )
+        )
+
+    def build_revision_url(self, build):
+        """
+        Build a Phabricator frontend url for a build's revision
+        """
+        return f"https://{self.api.hostname}/D{build.revision_id}"
+
+    def is_expired_build(self, build):
+        """
+        Check if a build has expired, using its Phabricator diff information
+        Returns True when the build has expired and should not be processed
+        """
+        assert isinstance(build, PhabricatorBuild)
+
+        # We need Phabricator diff details to get the date
+        if build.diff is None:
+            try:
+                diffs = self.api.search_diffs(diff_id=build.diff_id)
+                if not diffs:
+                    raise Exception(f"Diff {build.diff_id} not found on Phabricator")
+                build.diff = diffs[0]
+            except Exception as e:
+                logger.warn("Failed to load diff", build=str(build), err=str(e))
+                return False
+
+        # Then we can check on the expiry date
+        date_created = build.diff.get("dateCreated")
+        if not date_created:
+            logger.warn("No creation date found", build=str(build))
+            return False
+
+        logger.info("Found diff creation date", build=str(build), created=date_created)
+
+        return datetime.now() - datetime.fromtimestamp(date_created) > self.build_expiry
```

Comment:
Should this be `datetime.now(timezone.utc)` (with `utcnow()` deprecated) to avoid different behavior when executed locally?

Reply:
Probably but this could be a followup ? (as this is a 1-to-1 port of the libmozevent code).

BTW This snippet would then probably be "promoted" to a method shared by both Phabricator & Github sources